### PR TITLE
DOPE-230: added type functions and tests

### PR DESCRIPTION
**Pull request by martinagallati-ergon**
Edited to add: The type function `toAtom` is not implemented, because we always want to know the type of the passed parameter, so it requires users to use the specific functions `toStr, toBool, toNumber`

### CODE REVIEW

File: core/src/test/kotlin/ch/ergon/dope/buildTest/TypeFunctionsTest.kt [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+    @Test
+    fun `should support ISARRAY as boolean comparison`() {
+        val expected = "SELECT * FROM `someBucket` WHERE ISARRAY(`numberArrayField`)"
+
+        val actual = create
+            .selectFrom(
+                someBucket(),
+            )
+            .where(
+                isArray(someNumberArrayField()),
```
Comment (pgruntz) [inline core/src/test/kotlin/ch/ergon/dope/buildTest/TypeFunctionsTest.kt:41]: I would go for the syntax
```
...where(someNumberArrayField().isArray())
...where(someNumberArrayField().isAtom().and(someNumberArrayField().isBoolean())
...where(someNumberArrayField().toBoolean().isNumber())
```

File: crystal-map-connector/src/main/kotlin/ch/ergon/dope/extension/type/typefunction/IsArray.kt [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+import ch.ergon.dope.resolvable.expression.unaliased.type.typefunction.isArray
+import ch.ergon.dope.toDopeType
+import com.schwarz.crystalapi.schema.CMField
+import com.schwarz.crystalapi.schema.CMList
+
+@JvmName("isArrayNumber")
+fun isArray(field: CMField<out Number>) = isArray(field.toDopeType())
+
+@JvmName("isArrayString")
+fun isArray(field: CMField<String>) = isArray(field.toDopeType())
```
Comment (pgruntz) [inline crystal-map-connector/src/main/kotlin/ch/ergon/dope/extension/type/typefunction/IsArray.kt:12]: a lot of unused methods in these files

File: core/src/main/kotlin/ch/ergon/dope/validtype/ValidType.kt [hunk 3]
```diff
@@ -4,11 +4,13 @@ sealed interface ValidType
 
 interface ComparableType : ValidType
 
-interface BooleanType : ValidType
+interface AtomType : ValidType
```
Comment (pgruntz) [inline core/src/main/kotlin/ch/ergon/dope/validtype/ValidType.kt:7]: please update the diagram

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/typefunction/ToArrayExpression.kt [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+            parameters = expressionDopeQuery.parameters,
+        )
+    }
+}
+
+@JvmName("atomToArray")
+fun <T : AtomType> toArray(expression: TypeExpression<T>) = ToArrayExpression(expression)
+
+class ArrayToArrayExpression<T : ValidType>(
+    private val expression: TypeExpression<ArrayType<T>>,
```
Comment (pgruntz) [inline core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/typefunction/ToArrayExpression.kt:26]: I very much doubt whether we want to offer this functionality in dope: When do we want to pass an array to the toArray-method? Is there a use case?

**Review by pgruntz — CHANGES_REQUESTED** [review]

**Comment by pgruntz** [comment]
What about the _toAtom_-method?

**Review by pgruntz — CHANGES_REQUESTED** [review]
What about the toAtom()-method?

File: core/src/main/kotlin/ch/ergon/dope/validtype/ValidType.kt (hunk 3 above)
Comment (jansigi) [reply core/src/main/kotlin/ch/ergon/dope/validtype/ValidType.kt:7]: Why did you change that?

**Comment by martinagallati-ergon** [comment]
> What about the toAtom()-method?

@pgruntz In my notes I wrote that we talked about this offline and came to the conclusion to not add this function (only the "explicit" functions toString, toNumber, toBoolean). Do you see a need for discussion again?
Otherwise, I will just add it to the description of the PR for better traceability.

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/typefunction/ToArrayExpression.kt (hunk 4 above)
Comment (martinagallati-ergon) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/typefunction/ToArrayExpression.kt:26]: I can remove it no problem. We can always add it again if need be.

File: core/src/test/kotlin/ch/ergon/dope/buildTest/TypeFunctionsTest.kt (hunk 1 above)
Comment (martinagallati-ergon) [reply core/src/test/kotlin/ch/ergon/dope/buildTest/TypeFunctionsTest.kt:41]: I added this syntax for all `is...` and `to...` extensions. I left `typeOf(...)` as is for a more intuitive usage.

File: core/src/main/kotlin/ch/ergon/dope/validtype/ValidType.kt (hunk 3 above)
Comment (martinagallati-ergon) [reply core/src/main/kotlin/ch/ergon/dope/validtype/ValidType.kt:7]: @jansigi I changed it specifically for the `ToArrayExpression`, which behaves differently for atomic types (number, string, boolean). After a quick discussion with pgruntz I added the new interface 🙂

Comment (martinagallati-ergon) [reply core/src/main/kotlin/ch/ergon/dope/validtype/ValidType.kt:7]: > please update the diagram

done

File: crystal-map-connector/src/main/kotlin/ch/ergon/dope/extension/type/typefunction/IsArray.kt (hunk 2 above)
Comment (martinagallati-ergon) [reply crystal-map-connector/src/main/kotlin/ch/ergon/dope/extension/type/typefunction/IsArray.kt:12]: Not anymore, I added tests for each of them

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/typefunction/ToNumberExpression.kt [hunk 5]
```diff
@@ -21,9 +22,20 @@ class ToNumberExpression<T : ValidType>(
...
 
-fun toNumber(
-    expression: TypeExpression<StringType>,
-    stringExpression: TypeExpression<StringType>,
-) = ToNumberExpression(expression, stringExpression)
+fun String.toNumber() = ToNumberExpression(this.toDopeType())
+
+fun Boolean.toNumber() = ToNumberExpression(this.toDopeType())
+
+fun TypeExpression<StringType>.toNumber(stringExpression: TypeExpression<StringType>) =
```
Comment (pgruntz) [inline core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/typefunction/ToNumberExpression.kt:31]: can we name the parameter to something more meanigful? Like _filterChars_ or _removeChars_?

**Comment by pgruntz** [comment]
> > What about the toAtom()-method?
> 
> @pgruntz In my notes I wrote that we talked about this offline and came to the conclusion to not add this function (only the "explicit" functions toString, toNumber, toBoolean). Do you see a need for discussion again? Otherwise, I will just add it to the description of the PR for better traceability.

Ah yes I remeber, because the toAtom()-method basically just returns the parameter right?

**Review by pgruntz — APPROVED** [review]

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/typefunction/ToNumberExpression.kt (hunk 5 above)
Comment (martinagallati-ergon) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/typefunction/ToNumberExpression.kt:31]: I like _filterChars_ 🙂 I renamed it

**Pull request merged**